File: src/main/java/me/testaccount666/serversystem/commands/executables/smelt/CommandSmelt.java
package me.testaccount666.serversystem.commands.executables.smelt;

import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.ConsoleUser;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;
import org.bukkit.inventory.FurnaceRecipe;
import org.bukkit.inventory.ItemStack;

import java.util.Optional;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "smelt")
public class CommandSmelt extends AbstractServerSystemCommand {

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (!checkBasePermission(commandSender, "Smelt.Use")) return;
        if (commandSender instanceof ConsoleUser) {
            general("NotPlayer", commandSender).build();
            return;
        }

        var itemInHand = commandSender.getPlayer().getInventory().getItemInMainHand();
        if (itemInHand.getType().isAir()) {
            command("Smelt.NoItemInHand", commandSender).build();
            return;
        }

        var smeltedItemOptional = getSmeltedItem(itemInHand);
        if (smeltedItemOptional.isEmpty()) {
            command("Smelt.NoSmeltedItem", commandSender).build();
            return;
        }

        var smeltedItem = smeltedItemOptional.get();
        smeltedItem.setAmount(itemInHand.getAmount());
        commandSender.getPlayer().getInventory().setItemInMainHand(smeltedItem);

        command("Smelt.Success", commandSender).
                postModifier(message -> message.replace("<INPUT>", itemInHand.getType().name())
                        .replace("<OUTPUT>", smeltedItem.getType().name())).build();
    }

    private Optional<ItemStack> getSmeltedItem(ItemStack itemStack) {
        var recipeIterator = Bukkit.recipeIterator();

        while (recipeIterator.hasNext()) {
            var recipe = recipeIterator.next();
            if (!(recipe instanceof FurnaceRecipe furnaceRecipe)) continue;
            if (!furnaceRecipe.getInputChoice().test(itemStack)) continue;

            return Optional.of(furnaceRecipe.getResult());
        }

        return Optional.empty();
    }

    @Override
    public String getSyntaxPath(Command command) {
        throw new UnsupportedOperationException("Smelt command doesn't have an available syntax!");
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        return PermissionManager.hasCommandPermission(player, "Smelt.Use", false);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/smelt/CommandSmelt.kt
package me.testaccount666.serversystem.commands.executables.smelt

import me.testaccount666.paperktx.extensions.*
import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.commandMsg
import me.testaccount666.serversystem.userdata.User
import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.inventory.*

@ServerSystemCommand("smelt")
class CommandSmelt : AbstractServerSystemCommand() {
    override fun getUsagePermission(command: Command) = "Smelt.Use"
    override fun getSyntaxPath(command: Command?) = "Generic"

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        if (!isPlayer(commandSender)) return

        val player = commandSender.getPlayer()!!
        val itemInHand = player.inventory[EquipmentSlot.HAND]
        if (itemInHand.isAir()) {
            commandSender.commandMsg("Smelt.NoItemInHand")
            return
        }

        val smeltedItem = getSmeltedItem(itemInHand) ?: run {
            commandSender.commandMsg("Smelt.NoSmeltedItem")
            return
        }

        smeltedItem.amount = itemInHand.amount
        player.inventory[EquipmentSlot.HAND] = (smeltedItem)

        commandSender.commandMsg("Smelt.Success") {
            postModifier {
                it.replace("<INPUT>", itemInHand.type.name)
                    .replace("<OUTPUT>", smeltedItem.type.name)
            }
        }
    }

    private fun getSmeltedItem(itemStack: ItemStack): ItemStack? {
        return Bukkit.recipeIterator().asSequence().filterIsInstance<FurnaceRecipe>()
            .filter { it.inputChoice.test(itemStack) }.map(Recipe::getResult).firstOrNull()
    }
}
